File: W_3_Stack/Stack cpp/LStack.cpp
template <typename E> class LStack {

private:
    Link<E> *top;
    int size;
    static const int defaultSize = 10;


public:
    LStack(int sz = defaultSize) {
        top = NULL; //NULL
        size = 0;
    }

    ~LStack() {
        clear();
    }

    void clear() {
        while(top != NULL) {
            Link<E> *temp = top;
            top = top->next;
            delete temp;
        }

        size = 0;
     }





};

File: W_3_Stack/Java Stack/src/LStack.java
public class LStack<E> implements Stack<E>{
    private  Link<E> top;
    private int size;

    public LStack() {
        top = null;
        size = 0;
    }

    public LStack(int size) {
        this();
    }

    @Override
    public void clear() {
        top = null;
        size = 0;
    }

    @Override
    public void push(E it) {
        top = new Link<E> (it, top);
        size++;
    }

    @Override
    public E pop() {
        assert top != null: "Stack is empty";
        E it = top.element();
        top = top.next();
        size--;

        return it;
    }

    @Override
    public E topValue() {
        assert top != null: "Stack is empty";
        return top.element();
    }

    @Override
    public int length() {
        return size;
    }
}
